presto-tests/src/test/java/com/facebook/presto/execution/resourceGroups/db/TestQueuesDb.java

```diff
@@ -164,28 +172,31 @@ public void testTooManyQueries()
         QueryId thirdDashboardQuery = createQuery(queryRunner, dashboardSession(), LONG_LASTING_QUERY);
         waitForQueryState(queryRunner, thirdDashboardQuery, FAILED);
 
+        InternalResourceGroupManager manager = queryRunner.getCoordinator().getResourceGroupManager().get();
+        DbResourceGroupConfigurationManager dbConfigurationManager = (DbResourceGroupConfigurationManager) manager.getConfigurationManager();
+
         // Allow one more query to run and resubmit third query
         dao.updateResourceGroup(3, "user-${USER}", "1MB", 3, 4, 4, null, null, null, null, null, 1L, TEST_ENVIRONMENT);
         dao.updateResourceGroup(5, "dashboard-${USER}", "1MB", 1, 2, 2, null, null, null, null, null, 3L, TEST_ENVIRONMENT);
 
-        InternalResourceGroupManager manager = queryRunner.getCoordinator().getResourceGroupManager().get();
-        DbResourceGroupConfigurationManager dbConfigurationManager = (DbResourceGroupConfigurationManager) manager.getConfigurationManager();
-
         // Trigger reload to make the test more deterministic
         dbConfigurationManager.load();
-        waitForQueryState(queryRunner, secondDashboardQuery, RUNNING);
+        manager.refreshInternalResourceGroups();
+
+        waitForQueryState(queryRunner, secondDashboardQuery, QUEUED);
         thirdDashboardQuery = createQuery(queryRunner, dashboardSession(), LONG_LASTING_QUERY);
-        waitForQueryState(queryRunner, thirdDashboardQuery, QUEUED);
+        waitForQueryState(queryRunner, thirdDashboardQuery, RUNNING);
 
         // Lower running queries in dashboard resource groups and reload the config
         dao.updateResourceGroup(5, "dashboard-${USER}", "1MB", 1, 1, 1, null, null, null, null, null, 3L, TEST_ENVIRONMENT);
         dbConfigurationManager.load();
+        manager.refreshInternalResourceGroups();
 
         // Cancel query and verify that third query is still queued
         cancelQuery(queryRunner, firstDashboardQuery);
         waitForQueryState(queryRunner, firstDashboardQuery, FAILED);
         MILLISECONDS.sleep(2000);
-        waitForQueryState(queryRunner, thirdDashboardQuery, QUEUED);
+        waitForQueryState(queryRunner, thirdDashboardQuery, RUNNING);
```

Comment:
We should not have any change that would make existing tests fail right? If that's the case, please do not modify existing tests, but add new ones, if necessary.

Reply:
There is some changes in the way resource group will behave after introducing versioning:
 If a given resource group has a queued queries and we modify the resource group, a new version is created. And any new query submitted will be against the new resource group and can start executing immediately if it satisfies the hard concurrency. And that's what is happening in this case. 
Also Queries already queued in old resource group will still be in queue till a free slot is available within the old version. If we want to change that, we can move all the queued query from old resource group to new resource group. But that means we may end up having a resource group state where intermediate group may have running queries which we don't want. 
Let me know if that make sense or we want different behavior.

Reply:
Shouldn't it run the queued queries from the old version first before starting running queries from the new version?

Reply:
For any given resource group if it has multiple versions:
1. Old version try to run query based on it's resource setting. If it can run query, it will start the query.
2. If the old version can't run any more queries, we will check if it's newer version can run query based on new resource settings. 
For example, If version V1 of resource group has hard concurrency of 2. And version V2 of resource group has hard concurrency of 4. When resource group manager try to run query, it checks with V1 if it can run query, and it will allow V1 to run max 2 queries to run. Post that when it checks that V1 can't run more query, it will check V2 and allow it to run max of 2 queries given older version already running 2 queries. 
This is the new behavior of the resource groups and that causes some of the test cases to fail that needs to be modified. I am adding more tests case to cover this behavior as well.

Reply:
One decision is still open here:
What should we do with the already queued queries?
- In existing behavior: those queued queries will be run according to the new settings. 
- But in new behavior: queued queries remain with the old version of the resource group and executed as per the old settings. So if concurrency has increased for the resource group, old queued queries still run with the old concurrency and won't execute faster. Only newly submitted queries will get the new resource group settings.

I believe, that's a reasonable choice but want to make sure that's acceptable of we should move those queued queries against new version? One concern with this is: new version might not be a leaf resource group and so we may end up having internal node having queued queries.

Reply:
Consider the following scenario:
Some high pri resource group is queued up and they need more capacity urgently. We can do that by changing the resource configuration and allow more queries to run. However, in the current approach, these queries will still be queued. The only way to make them running is to kill them and retry.

I think it makes more sense to first drain the queue from the old version. So maybe a better thing to do is: if both versions are leaf, just move the queued queries from old version to new.

For resource groups that changed from leaf to internal, it makes more sense to finish scheduling all the running and queued queries of the old version before switching to the new version (so new version would only queue queries but not running them).

The down side of this is how versions are handled is not very consistent. So another idea would be to build a new version of the whole tree rather than updating nodes. We briefly discussed this earlier. The downside of this approach is that if you wait until the previous tree is drained before switching, this might take a long time and waste resources. One way to address this is to take all queries from the old version, and use the new selector rules to assign them to the new version (first go through all running queries and ignore capacity constraint, because they are already running, then go through all queued queries and apply normal rules).

Reply:
done the changes as suggested and also updated description explaining the behavior of the resource group versioning. Please review it and let me know if anything is needed.